File: Semester 03 Projects/Solitair Game/BL/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Solitair_Game
{
    //Node Class Containing a Card and it's next node object.
    public class Node
    {
        public Card CurrentCard;
        public Node Next;
    }

    //Stack Class implementing stack using Linked List.
    public class Stack
    {
        public Node Top;
        public Stack()
        {
            Top = null;
        }


        //Function for adding card into stack.
        public void Push(Card newcard)
        {
            Node newNode = new Node();
            newNode.CurrentCard = newcard;
            newNode.Next = Top;
            Top = newNode;

        }
        //Function for removing card from stack.
        public Card Pop()
        {
            if (IsEmpty())
            {
                return null;
            }
            Node next = Top.Next;
            Card CardToDelete = Top.CurrentCard;
            Top = next;
            return CardToDelete;
        }
        //Function for returning top of stack.
        public Card Peek()
        {
            if (IsEmpty())
            {
                return null;
            }
            return Top.CurrentCard;
        }
        //Function for checking stack is empty or not.
        public bool IsEmpty()
        {
            if(Top == null)
            {
                return true;
            }
            return false;
        }
        //Function for counting number of cards in stack.
        public int GetTotalNumberOfCards()
        {
            int count = 0;
            Node current = Top;
            while (current != null)
            {
                count++;
                current = current.Next;
            }
            return count;
        }
        //Function for making Deep Copy of Stack.
        public Stack DeepCopy()
        {
            Stack copy = new Stack();
            if (Top == null) 
             return copy;

            Node current = Top;
            Stack tempStack = new Stack();

            while (current != null)
            {
                if (current.CurrentCard == null)
                    break;
                Card copycard = new Card(current.CurrentCard);
                tempStack.Push(copycard);
                current = current.Next;
            }

            current = tempStack.Top;
            while (current != null)
            {
                copy.Push(current.CurrentCard);
                current = current.Next;
            }

            return copy;
        }

    }

}

File: DSA-Final-Project/Server/DataStructuresAndAlgorithms/Stack.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.head = null;
    }

    Push(value) {
        const newNode = new Node(value);
        newNode.next = this.head;
        this.head = newNode;
    }

    Pop() {
        if (this.IsEmpty()) {
            return "No elements to pop";
        }
        const toDelete = this.head.value;
        this.head = this.head.next;
        return toDelete;
    }

    Peek() {
        if (this.IsEmpty()) {
            return "Stack is empty";
        }
        return this.head.value;
    }

    IsEmpty() {
        return this.head === null;
    }
    DeleteSpecific(location) {
        if (this.head == null) {
            return false;
        }

        if (this.head.value.UserID === location.UserID &&
            this.head.value.SourceLocation === location.SourceLocation &&
            this.head.value.DestinationLocation === location.DestinationLocation) 
        {
            const locationToDelete = this.head;
            this.head = this.head.next; 
            return locationToDelete.value;
        }

        let current = this.head;
        while (current.next != null) 
            {
            if (current.next.value.UserID === location.UserID &&
                current.next.value.SourceLocation === location.SourceLocation &&
                current.next.value.DestinationLocation === location.DestinationLocation)
                {
                const locationToDelete = current.next;
                if (current.next.next != null) 
                {
                    current.next = current.next.next; 
                } 
                else 
                {
                    current.next = null; 
                }
                return locationToDelete.value;
            }
            current = current.next;
        }
    
        return false; 
    }
}

module.exports = Stack;

